File: src/AAS.BackOffice/client_app/src/domain/emailVerifications/emailVerificationsProvider.ts
import HttpClient from "../../tools/httpClient";
import { Result, mapToResult } from "../../tools/types/results/result";
import { User, mapToUser } from "../users/user";
import { EmailVerification, mapToEmailVerification } from "./emailVerification";

export namespace EmailVerificationsProvider {
    export async function resendEmailVerificationMessage(userEmail: string | null): Promise<Result> {
        const result = await HttpClient.getJsonAsync("/email_verificaton/resend_email_verification_message", { userEmail });
        return mapToResult(result);
    }

    export async function getUserEmailVerification(userEmailVerificationToken: string): Promise<{ user: User, emailVerificaton: EmailVerification } | null> {
        const details = await HttpClient.getJsonAsync("/email_verifications/get_user_email_verification", { userEmailVerificationToken });

        if (details === null) return null;

        return {
            user: mapToUser(details.user),
            emailVerificaton: mapToEmailVerification(details.emailVerification)
        };
    }

    export async function confirmEmail(userEmailVerificationToken: string | null): Promise<Result> {
        const result = await HttpClient.getJsonAsync("/email_verifications/confirm_email", { userEmailVerificationToken });
        return mapToResult(result);
    }
}

File: src/AAS.Services/EmailVerifications/Providers/EmailVerificationsProvider.cs
using AAS.Configurator;
using AAS.Domain.EmailVerifications;
using AAS.Domain.Users;
using AAS.Tools.Types.Results;
using System.Net;
using System.Net.Mail;

namespace AAS.Services.EmailVerifications.Providers;

public class EmailVerificationsProvider : IEmailVerificationsProvider
{
    public Result SendVerificationMessage(EmailVerification emailVerification, User userForVerification)
    {
        try
        {
            string subject = "Подтверждение Email";
            string emailVerificationUrl = Configurations.EmailVerification.VerificationUrlTemplate.Replace(
                $"{Configurations.EmailVerification.VerificationUrlReplacingValue}", emailVerification.Token
            );
            MailMessage message = new(Configurations.EmailVerification.Login, userForVerification.Email, subject, null);
            message.IsBodyHtml = true;
            message.Body += $"Перейдите по следующей ссылке для подтверждения своей почты<br/><a href='{emailVerificationUrl}'> Подтвердить.</a>";

            using SmtpClient client = new("smtp.gmail.com", 587);

            client.EnableSsl = true;
            client.UseDefaultCredentials = false;
            client.Credentials = new NetworkCredential(Configurations.EmailVerification.Login, Configurations.EmailVerification.Password);

            client.Send(message);

            return Result.Success();
        }
        catch
        {
            return Result.Fail("Не удалось отправить сообщение с подтверждением email адреса");
        }
    }
}
